src/DotVVM.Framework/Resources/Scripts/collections/arrayHelper.ts

```diff
@@ -1,121 +1,215 @@
 ﻿import { orderBy, orderByDesc } from './sortingHelper'
 
 export {
+    add,
+    addOrUpdate,
+    addRange,
     all,
     any,
     clear,
     distinct,
     firstOrDefault,
-    forEach,
+    insert,
+    insertRange,
     lastOrDefault,
     max,
     min,
     orderBy,
     orderByDesc,
-    remove,
-    removeFirst
+    removeAll,
+    removeAt,
+    removeFirst,
+    removeLast,
+    removeRange,
+    reverse
 }
 
-function any<T>(source: T[], predicate: (s: T) => boolean): boolean {
-    return firstOrDefault(source, predicate) != null;
+function add<T>(observable: any, element: T): void {
+    let array = [...observable.state, element];
+    observable.setState(array);
 }
 
-function all<T>(source: T[], predicate: (s: T) => boolean) {
-    for (let i = 0; i < source.length; i++) {
-        if (!predicate(source[i])) {
+function addOrUpdate<T>(observable: any, element: T, matcher: (e: T) => boolean, updater: (e: T) => T): void {
+    let array = Array.from<T>(observable.state);
+    let found = false;
+
+    for (let i = 0; i < array.length; i++) {
+        if (!matcher(array[i])) {
+            continue;
+        }
+
+        found = true;
+        array[i] = updater(array[i]);
+    }
+
+    if (!found) {
+        array.push(element);
+    }
+
+    observable.setState(array);
+}
+
+function addRange<T>(observable: any, elements: T[]): void {
+    let array = Array.from<T>(observable.state);
+    for (let i = 0; i < elements.length; i++) {
+        array.push(ko.unwrap(elements[i]));
+    }
+
+    observable.setState(array);
+}
+
+function any<T>(array: T[], predicate: (s: T) => boolean): boolean {
+    return firstOrDefault(array, predicate) != null;
+}
+
+function all<T>(array: T[], predicate: (s: T) => boolean): boolean {
+    for (let i = 0; i < array.length; i++) {
+        if (!predicate(array[i])) {
             return false;
         }
     }
     return true;
 }
 
-function clear<T>(source: T[]): void {
-    source.splice(0, source.length);
+function clear(observable: any): void {
+    observable.setState([]);
 }
 
-function distinct<T>(source: T[]): T[] {
+function distinct<T>(array: T[]): T[] {
     let r = [];
-    for (let i = 0; i < source.length; i++) {
+    for (let i = 0; i < array.length; i++) {
         let found = false;
         for (var j = 0; j < r.length; j++) {
-            if (r[j] == source[i]) {
+            if (r[j] == array[i]) {
                 found = true;
                 break;
             }
         }
         if (found)
             continue;
-        r.push(source[i]);
+        r.push(array[i]);
```

Comment:
you can do just `return Array.from(new Set(array))`, maybe `return Array.from(new Set(array.map(ko.unwrap)))`, not sure.

Reply:
The second option works